### ISSUE TITLE
It is not possible to send `null` as the last argument of an UndoRedo method

### ISSUE THREAD

**Issue by Zylann** [godotengine/godot#36895]
Godot 3.2

To register a method to be executed when doing or undoing a change, `UndoRedo.add_do_method()` or `UndoRedo.add_undo_method()` have to be used.
Those methods were implemented by declaring 5 `Variant` arguments, all `null` by default.
Unfortunately, the argument count is then determined by checking how many remaining arguments are null, *including cases where you want to call the method with `null`*.
That means it is impossible to write this:
```gdscript
undo_redo.add_do_method(object, "foo", 42, true, null)
```
Because the `null` at the end will be confused as absence of argument. Then when committing the action, undoing or redoing it, it won't work and this kind of error will be printed:
```
ERROR: UndoRedo::_process_operation_list: Error calling method from signal 'foo': Method expected 3 arguments, but called with 2.
```
See https://github.com/godotengine/godot/blob/78adce833bce49ef1bf1503956a6162a1c6dd8b2/core/undo_redo.cpp#L284

I had to workaround this limitation by binding extra methods in my objects, and check if variants are null before using `add_do_method` and `add_undo_method`.

To fix this, `UndoRedo` should not determine argument count in this way. Instead, it should use reflection to query how many arguments the called method actually has. Other solutions include sending an array explicitely, adding a count argument, or having a `Variant::UNDEFINED` type, but that would break compatibility.

**Comment by KoBeWi** [comment]
Related to #29204

**Comment by kopcion** [comment]
I've been working on this bug. For now, I only managed to solve this issue by having map in UndoRedo class with argc for current functions, but I would like to do some more, so feedback would be great. I don't see how to apply reflection to this problem, nor what exactly do you mean by reflection in c++ (until recently it wasn't included afaik).

Here is basic project reproducing the bug. 
[Issue#36895.zip](https://github.com/godotengine/godot/files/4499973/Issue.36895.zip)

**Comment by Jummit** [comment]
[undo_test.zip](https://github.com/godotengine/godot/files/5447322/undo_test.zip)

Minimal reproduction project I made for #43141

**Comment by me2beats** [comment]
just faced this too, so I do it in my own functions like
```gdscript
undo.add_do_method(self, "_do")
```

**Comment by akien-mga** [comment]
Fixed by #66070.

**Comment by KoBeWi** [comment]
This was actually resolved by #58929
UndoRedo uses variadic functions now and has no limit.
